mmv1/third_party/terraform/services/certificatemanager/resource_certificate_manager_dns_authorization_upgrade_test.go

```diff
@@ -0,0 +1,77 @@
+package certificatemanager_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+)
+
+// Tests schema version migration by creating a dns authorization with an old version of the provider (5.15.0)
+// and then updating it with the current version the provider.
+func TestAccCertificateManagerDnsAuthorization_migration(t *testing.T) {
+	acctest.SkipIfVcr(t)
+	t.Parallel()
+	name := fmt.Sprintf("tf-test-%d", acctest.RandInt(t))
+
+	oldVersion := map[string]resource.ExternalProvider{
+		"google": {
+			VersionConstraint: "5.15.0", // a version that doesn't support location yet.
+			Source:            "registry.terraform.io/hashicorp/google",
+		},
+	}
+	newVersion := map[string]func() (*schema.Provider, error){
+		"mynewprovider": func() (*schema.Provider, error) { return acctest.TestAccProviders["google"], nil },
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:     func() { acctest.AccTestPreCheck(t) },
+		CheckDestroy: testAccCheckCertificateManagerDnsAuthorizationDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config:            dnsAuthorizationResourceConfig(name),
+				ExternalProviders: oldVersion,
+			},
+			{
+				ResourceName:            "google_certificate_manager_dns_authorization.default",
+				ImportState:             true,
+				ImportStateVerifyIgnore: []string{"location"},
+				ExternalProviders:       oldVersion,
+			},
+			{
+				Config:            dnsAuthorizationResourceConfigUpdated(name),
+				ProviderFactories: newVersion,
+			},
+			{
+				ResourceName:            "google_certificate_manager_dns_authorization.default",
+				ImportState:             true,
+				ImportStateVerifyIgnore: []string{"location"},
+				ProviderFactories:       newVersion,
+			},
+		},
+	})
+}
+
+func dnsAuthorizationResourceConfig(name string) string {
+	return fmt.Sprintf(`
+	resource "google_certificate_manager_dns_authorization" "default" {
+		name        = "%s"
+		description = "The default dns"
+		domain      = "domain.hashicorptest.com"
+	  }
+	`, name)
+}
+
+func dnsAuthorizationResourceConfigUpdated(name string) string {
+	return fmt.Sprintf(`
+	provider "mynewprovider" {}
+	
+	resource "google_certificate_manager_dns_authorization" "default" {
+		name        = "%s"
+		description = "The migrated default dns"
+		domain      = "domain.hashicorptest.com"
+	  }
+	`, name)
+}
```

Comment:
To enable you to see the debug log mentioned in another comment, you either need to change this code to:

```diff
resource "google_certificate_manager_dns_authorization" "default" {
+ 		provider = mynewprovider
		name        = "%s"
		description = "The migrated default dns"
		domain      = "domain.hashicorptest.com"
	  }
```

or remove the `provider "mynewprovider" {}` line and change `newVersion ` to have a `"google` key instead of `"mynewprovider"`

Reply:
To follow this up, without the explicit `provider` argument Terraform will use the `google_` prefix of the resource to decide which provider is needed. I'm not sure exactly what is happening, but it may be that the "google" provider from `oldVersion` is still accessible and being used 😱 . If we change `newVersion` to have a "google" key we would be ensuring the expected code is used. Adding `provider = mynewprovider` would solve the problem in a different way because it stops Terraform looking for "google" and instead looks for "mynewprovider".

Reply:
Thanks for the thorough explanation and for testing this locally! I really appreciate the efforts!

I have tried to do as mentioned, but I still can not see the print statement in the debug logs (I made sure that TF_LOG=DEBUG). 
I first tried to remove mynewprovoider at from the example, and replace "mynewprovider" key in newVersion with "google".  The second thing I tried was to embed `provider = mynewprovider` inside the updated example as you suggested. 

Did I miss something?

Reply:
@SarahFrench 
I also tried to run an existing upgrade test:`make testacc TEST=./google/services/cloudbuild TESTARGS='-run=TestAccCloudBuildTrigger_migration'
` and it's not printing the debug logs as well (https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/state_migrations/cloud_build_trigger.go.erb#L2)

Reply:
Do you have TF_LOG=DEBUG set to make those logs visible? If you run `TF_LOG=DEBUG make testacc TEST=./google/services/cloudbuild TESTARGS='-run=TestAccCloudBuildTrigger_migration'` you should be able to see the 'Attributes before migration..." "Attributes after migration..." log entries. I can see that debug log when I run it locally, and it works because that test uses "mynewprovider" and [has that explicitly set as the provider to be used for the resource](https://github.com/hashicorp/terraform-provider-google/blob/0b64b416469014808ba1e27b6de9e0a9af473d47/google/services/cloudbuild/resource_cloudbuild_trigger_upgrade_test.go#L81) in the relevant test step.

Reply:
@SarahFrench 
Yes, I executed `TF_LOG=DEBUG make testacc TEST=./google/services/cloudbuild TESTARGS='-run=TestAccCloudBuildTrigger_migration'` and can not see the log 'Attributes before migration...'

This is the logs I receive: https://paste.googleplex.com/5093583482257408

Reply:
I'm afraid I'm not a Googler so I don't have access to that resource, and I don't have a good alternative for privately sharing logs that might contain sensitive info.

I ran the certificatemanager package's acceptance tests in our overnight test environment using [this PR's proposed code changes in the GA provider repo](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9968-old..auto-pr-9968) and I can see this in the debug logs for `TestAccCertificateManagerDnsAuthorization_migration`:

```
2024/02/21 14:36:37 [DEBUG] Attributes before migration: map[string]interface {}{"description":"The default dns", "dns_resource_record":[]interface {}{map[string]interface {}{"data":"3ee21668-b128-4d5d-9c72-34e59dd32c04.13.authorize.certificatemanager.goog.", "name":"_acme-challenge.domain.hashicorptest.com.", "type":"CNAME"}}, "domain":"domain.hashicorptest.com", "effective_labels":map[string]interface {}{}, "id":"projects/ci-test-project-188019/locations/global/dnsAuthorizations/tf-test-8492830205563058242", "labels":map[string]interface {}{}, "name":"tf-test-8492830205563058242", "project":"ci-test-project-188019", "terraform_labels":map[string]interface {}{}, "timeouts":interface {}(nil)}
2024/02/21 14:36:37 [DEBUG] Attributes after migration: map[string]interface {}{"description":"The default dns", "dns_resource_record":[]interface {}{map[string]interface {}{"data":"3ee21668-b128-4d5d-9c72-34e59dd32c04.13.authorize.certificatemanager.goog.", "name":"_acme-challenge.domain.hashicorptest.com.", "type":"CNAME"}}, "domain":"domain.hashicorptest.com", "effective_labels":map[string]interface {}{}, "id":"projects/ci-test-project-188019/locations/global/dnsAuthorizations/tf-test-8492830205563058242", "labels":map[string]interface {}{}, "location":"global", "name":"tf-test-8492830205563058242", "project":"ci-test-project-188019", "terraform_labels":map[string]interface {}{}, "timeouts":interface {}(nil)}
```
---

If you have TF_LOG=DEBUG set to see detailed logs, TF_ACC=1 set to enable acceptance tests running (set for you via `make testacc`), and you're using the correct code that includes the new TestAccCertificateManagerDnsAuthorization_migration test I'm not sure what the problem would be. As you seem to be a Googler, it might be worth attending the Google provider team's internal office hours to ask for help.

Reply:
Thanks again for helping testing this! 

From the logs you shared, the `location` field seems to exist after the migration. Should anything else be done from my side?

Reply:
Nothing more - thanks for your patience!